python/sglang/srt/reasoning_parser.py

```diff
@@ -0,0 +1,187 @@
+import re
+from typing import Optional, Dict
+
+REASONING_MODELS = ["deepseek-r1"]
+
+
+def is_reasoning_model(model_name: str) -> bool:
+    """Checks if the model is a reasoning model."""
+    return model_name.lower() in REASONING_MODELS
+
+
+class StreamingParseResult:
+    """Result of streaming incremental parsing."""
+    def __init__(self, normal_text: str = "", reasoning_text: str = ""):
+        self.normal_text = normal_text
+        self.reasoning_text = reasoning_text
+
+class BaseReasoningFormatDetector:
+    """Base class providing two sets of interfaces: one-time and streaming incremental."""
+    def __init__(self, stream_reasoning: bool = False):
+        self._buffer = ""
+        self._in_reasoning = False
+        self._current_reasoning = ""
+        self.stream_reasoning = stream_reasoning
+
+    def detect_and_parse(self, text: str) -> StreamingParseResult:
+        """Parses the text in one go."""
+        raise NotImplementedError
+
+    def parse_streaming_increment(self, new_text: str) -> StreamingParseResult:
+        """Streaming incremental parsing."""
+        raise NotImplementedError
+
+class DeepSeekR1Detector(BaseReasoningFormatDetector):
+    """
+    Detector for DeepSeek-R1 model.
+    Assumes reasoning format:
+      <think>...</think>
+    Returns all the text within the <think> tags as `reasoning_text` 
+    and the rest of the text as `normal_text`.
+
+    Args:
+        stream_reasoning (bool): If False, accumulates reasoning content until the end tag.
+            If True, streams reasoning content as it arrives.
+    """
+    def __init__(self, stream_reasoning: bool = False):
+        super().__init__(stream_reasoning=stream_reasoning)
+        self.think_start_token = "<think>"
+        self.think_end_token = "</think>"
+        self.reasoning_regex = re.compile(
+            rf"{self.think_start_token}(.*?){self.think_end_token}", 
```

Comment:
https://huggingface.co/deepseek-ai/DeepSeek-R1/commit/8a58a132790c9935686eb97f042afa8013451c9f updated the chat template, there is no start token now.

Reply:
Interesting, thanks for the heads up.
I don't really like the regex approach anyway since it doesn't capture truncated thinking (i.e when max_tokens is hit before </think>).

I suppose the model is assumed to be thinking until </think> is seen? I wonder if there's time where it won't output it?

Reply:
I implemented the feature in vLLM using regex, similar to your code. You can see it [here](https://github.com/vllm-project/vllm/pull/12473/files#diff-c4b04f75adef67f654db6708652a77a0d5ccc6e73a9c7a442abf134cf84fc266R30).

I’m not a fan of regex either, but I don’t have a better solution at the moment.

> I suppose the model is assumed to be thinking until is seen

Yes, I think so. Until </think> presents.

Reply:
Pushing updates soon.